services/src/main/java/org/keycloak/services/clientpolicy/executor/SecureRedirectUrisEnforcerExecutor.java

```diff
@@ -0,0 +1,675 @@
+                }
+                // verify and match redirect_uri parameter with registered redirect URIs
+                if (!isMatchedRedirectUris(redirectUriParam, client.getRedirectUris())) {
+                    throw invalidRedirectUri(ERR_AUTHZPARAM);
+                }
+                return;
+            default:
+                return;
+        }
+    }
+
+    void verifyRedirectUris(List<String> redirectUris) throws ClientPolicyException {
+        // open redirect allows any value for redirect uri
+        if (configuration.isAllowOpenRedirect()) {
+            if (!"dev".equals(System.getProperty("kc.profile"))) {
+                throw invalidRedirectUri(ERR_OPENREDIRECT);
+            }
+            return;
+        }
+
+        for (String uri : redirectUris) {
+            verifyRedirectUri(uri, false);
+        }
+    }
+
+    void verifyRedirectUri(String redirectUri, boolean isRedirectUriParam) throws ClientPolicyException {
+        UriValidation validation = null;
+
+        try {
+            validation = new UriValidation(redirectUri, isRedirectUriParam, configuration);
+        } catch (URISyntaxException e) {
+            logger.debugv("URISyntaxException - input = {0}, errMessage = {1], errReason = {2}", e.getInput(), e.getMessage(), e.getReason());
+            throw invalidRedirectUri(ERR_GENERAL);
+        }
+ 
+        validation.validate();
+    }
+
+    boolean isMatchedRedirectUris(String redirectUriParameter, Set<String> registeredRedirectUris) throws ClientPolicyException {
+        // open redirect allows any value for redirect uri
+        if (configuration.isAllowOpenRedirect()) {
+            if (!"dev".equals(System.getProperty("kc.profile"))) {
+                throw invalidRedirectUri(ERR_OPENREDIRECT);
+            }
+            return true;
+        }
+
+        if (redirectUriParameter == null || redirectUriParameter.isEmpty()) {
+            throw invalidRedirectUri(ERR_AUTHZPARAM);
+        }
+
+        try {
+            verifyRedirectUri(redirectUriParameter, true);
+        } catch (ClientPolicyException cpe) {
+            logger.debugv("exception in verifying redirect_uri with registered redirect uris. errorDetail = {0}", cpe.getErrorDetail());
+            throw invalidRedirectUri(ERR_AUTHZPARAM);
+        }
+
+        UriValidation validation = null;
+        UriMatching uriMatching = null;
+        UriType uriType = null;
+
+        try {
+            validation = new UriValidation(redirectUriParameter, true, configuration);
+            uriType = validation.identifyUriType();
+        } catch (URISyntaxException e) {
+            logger.debugv("URISyntaxException - input = {0}, errMessage = {1], errReason = {2}", e.getInput(), e.getMessage(), e.getReason());
+            throw invalidRedirectUri(ERR_AUTHZPARAM);
+        }
+
+        try {
+            uriMatching = new UriMatching(uriType, redirectUriParameter, registeredRedirectUris, configuration);
```

Comment:
I don't know if this check is necessary. The redirect URI will be validated against the valid redirect uris configured in the client later. I suposse this is just done to validate the passed redirect if the client has invalid valid redirects defined (because it was updated/created before applying the policy).

I would just check that the passed redirect URI in the authorization request is OK (the previous part `verifyRedirectUri(redirectUriParameter, true)`) and then just normal processing of the RedirectUtils. I would not mess with comparisons with the valid redirect URIs here when that is again done later. WDYT?

Reply:
Yes, I removed `UriMatching` logics.